### ISSUE TITLE
How to get output mask same size as input image

### ISSUE THREAD

**Issue by pnrl** [matterport/Mask_RCNN#2190]
Sir, I used 1536 X 2048 high resolution images and I changed Image resize mode as none and use min mask as false and assigned max and min sizes to overwrite config.py. I am getting output image as well as masks with size 288 X 432. I simply used visualize.py file only and in that I customized display_images to save images. What changes I need to made so that the output image and masks are of the same size as input image. In the attached I highlighted two functions where I make changes. I am simply using demo file in sample folders. Please help me.
[visualize.py.docx](https://github.com/matterport/Mask_RCNN/files/4639654/visualize.py.docx)

**Comment by suchiz** [comment]
Hi there, take a look at this thread. https://github.com/matterport/Mask_RCNN/issues/2177

**Comment by pnrl** [comment]
Sir,

Before I opened I checked the thread 2177 and I made those changes but I not able to get it so, I opened as new issue.

One question is how can I get the mask (I am new).

I made some few changes in visualize.py (display_top_masks & display_images) but I get the reduced sized mask instead of original size.

**Comment by suchiz** [comment]
I should have misunderstood what you are aiming for then.

You just want to save the binary mask (b&w) ?

**Comment by pnrl** [comment]
Yes Sir. From the original image, I am trying to extract individual objects as masked image. So that I will extract all the pixels of the object I am interested from the original image. So, can you help me to get the masked image as same as original image size.

**Comment by suchiz** [comment]
Alright, so first, do as it is said in the thread 2177, change square mode into none mode. Then, in utils.py, at unmold_mask function, save "full_mask" image just before returning it. It should do the job.

**Comment by pnrl** [comment]
Sorry to ask again. From where the unmold_mask will call?

**Comment by suchiz** [comment]
In detect function from model.py. Then visualize uses the results from "detect" function to draw them, but instead of drawing them, you can juste save them.

**Comment by pnrl** [comment]
Sir,

Thank you very much, by your help I could able to get the mask same size as my input image. Thank you very much.

Sir, I have another problem. From the same dataset images, the demo.ipymb is working for some images (input 1) and through an error for some other images (input 2). I found that the shape is giving as shape: (1536, 2048, 4)  and through an value error: ValueError: operands could not be broadcast together with shapes (1536,2048,4) (3,)

I didn’t understand why? Can you please help me.

Chandra Sekhar

 

 

 

Input 1:

 

Processing 1 images

 

image                    shape: (1536, 2048, 3)       min:    0.00000  max:  255.00000  uint8

molded_images            shape: (1, 1536, 2048, 3)    min: -123.70000  max:  151.10000  float64

image_metas              shape: (1, 93)               min:    0.00000  max: 2048.00000  int64

anchors                  shape: (1, 785664, 4)        min:   -0.23586  max:    1.19416  float32

 

 

Input 2:

 

Processing 1 images

image                    shape: (1536, 2048, 4)       min:    0.00000  max:  255.00000  uint8

 

ValueError                                Traceback (most recent call last)

<ipython-input-6-1164b49e6450> in <module>

      4 # Run detection

      5 #image = skimage.io.imread(os.path.join(IMAGE_DIR, 'splicing-16.png'))

----> 6 results = model.detect([image], verbose=1)

      7 

      8 

 

~/anaconda3/lib/python3.7/site-packages/mask_rcnn-2.1-py3.7.egg/mrcnn/model.py in detect(self, images, verbose)

   2501 

   2502         # Mold inputs to format expected by the neural network

-> 2503         molded_images, image_metas, windows = self.mold_inputs(images)

   2504 

   2505         # Validate image sizes

 

~/anaconda3/lib/python3.7/site-packages/mask_rcnn-2.1-py3.7.egg/mrcnn/model.py in mold_inputs(self, images)

   2400                 max_dim=self.config.IMAGE_MAX_DIM,

   2401                 mode=self.config.IMAGE_RESIZE_MODE)

-> 2402             molded_image = mold_image(molded_image, self.config)

   2403             # Build image_meta

   2404             image_meta = compose_image_meta(

 

~/anaconda3/lib/python3.7/site-packages/mask_rcnn-2.1-py3.7.egg/mrcnn/model.py in mold_image(images, config)

   2801     colors in RGB order.

   2802     """

-> 2803     return images.astype(np.float32) - config.MEAN_PIXEL

   2804 

   2805 

 

ValueError: operands could not be broadcast together with shapes (1536,2048,4) (3,)

**Comment by suchiz** [comment]
Hi there, you simply just cannot give images with more than 3 channels. As you may know, JPEG are just 3 channels (RGB), whereas PNG has 4 (RGBA). You also have the possibility to give 1 channel only (grayscale or binary) but then, check the wiki on how to proceed. What I just said to you is also mentionned in the wiki, so please, read it carefully, it can help you a lot ! :)

**Comment by pnrl** [comment]
Thank you Sir. I will check.

Another thing I found is that, the compressed masks exactly the same as the corresponding compressed image whereas when I use unmould_image, the masks I get are vary with the original image. It may be we expanded the masks. How to get the correct shape of the objects masks corresponding to original image.

**Comment by pnrl** [comment]
Thank you Sir. I will check.

Another thing I found is that, the compressed masks exactly the same as the
corresponding compressed image whereas when I use unmould_image, the masks
I get are vary with the original image. It may be we expanded the masks.
How to get the correct shape of the objects masks corresponding to original
image.

I also attached the sample segments for your reference.
With Thanks & Regards,
PNRL Chandra Sekhar
Associate Professor, Dept of CSE,
GIT, GITAM,
Visakhapatnam-45 (AP-INDIA)

**Comment by suchiz** [comment]
Hi there, 

Unfortunatly I'm using github web platform, I cannot receive your attachements.

Well, this is an active research field haha :D ! Not a simple question ! This is due to the way you interpolate your pixels while resizing (unmolding) your image. It is way easier to match perfectly on a small image because there's less pixels, hence less details. But it becomes harder when you have a high quality image. That's why it doesn't match as it gets bigger. You can try several methods of interpolation to see if you get a better result. But honestly, those people made MaskRCNN, one of the most effective CNN, I'd trust the function they chose to interpolate their soft mask :).

**Comment by pnrl** [comment]
Sir,
Sure, they does a great job. I too agree. It is admitted that I am using
high quality images. Thank you for your response
.With Thanks & Regards,
PNRL Chandra Sekhar
Associate Professor, Dept of CSE,
GIT, GITAM,
Visakhapatnam-45 (AP-INDIA)
